Update Chainnotifier Service to leverage batch-request support

Vib-UX opened the pull request:
BatchRPC

guggero commented on chainntnfs/bitcoindnotify/bitcoind.go:
```diff
@@ -42,6 +43,8 @@ type BitcoindNotifier struct {
 	chainConn   *chain.BitcoindClient
 	chainParams *chaincfg.Params
 
+	client *rpcclient.Client
```
nit: seems to be unused.

guggero commented on chainntnfs/bitcoindnotify/bitcoind.go:
```diff
@@ -780,22 +783,32 @@ func (b *BitcoindNotifier) historicalSpendDetails(
-	// was spent.
+	// batchrequest will store the scan requests at every height to determine
+	// if the output was spent
+	batchRequest := []rpcclient.FutureGetBlockHashResult{}
```
This could be a map with the height as the key. And we should pre-allocate as well to avoid it being resized (the number of elements is `endHeight - startHeight`).

guggero commented on chainntnfs/bitcoindnotify/bitcoind.go:
```diff
@@ -780,22 +783,32 @@ func (b *BitcoindNotifier) historicalSpendDetails(
+		}
+		batchRequest = append(batchRequest, sendReq)
+	}
+	// Sending the bulk request using updated batchClient
```
nit: add newline before comments and always use full sentences (with punctuation/full stops).

guggero commented on chainntnfs/bitcoindnotify/bitcoind.go:
```diff
@@ -780,22 +783,32 @@ func (b *BitcoindNotifier) historicalSpendDetails(
+	// Sending the bulk request using updated batchClient
+	b.chainConn.Send()
+
+	for height, response := range batchRequest {
```
With the slice this isn't really the `height` anymore but the index, which requires the change in the error message below. With the map (see comment above), this would become the `height` again.

Vib-UX replied:
Updated!

guggero reviewed (commented):
Looks pretty good! Added a few ideas for optimizations.

bhandras commented on chainntnfs/bitcoindnotify/bitcoind.go:
```diff
@@ -525,30 +526,44 @@ func (b *BitcoindNotifier) confDetailsManually(confRequest chainntnfs.ConfReques
-	// transaction was included in.
+	// batchRequest will store the scan requests at every height to determine
+	// where the transaction was included in.
+	batchRequest := make(map[uint32]rpcclient.FutureGetBlockHashResult, currentHeight-heightHint)
```
nit: formatting

bhandras commented on chainntnfs/bitcoindnotify/bitcoind.go:
```diff
@@ -525,30 +526,44 @@ func (b *BitcoindNotifier) confDetailsManually(confRequest chainntnfs.ConfReques
+		//BatchClient works with AsynRequest.
+		sendReq, err := b.chainConn.GetBlockHashAsync(int64(height))
+		if err != nil {
+			return nil, chainntnfs.TxNotFoundManually, fmt.Errorf("unable to receive "+
```
nit: formatting

bhandras commented on chainntnfs/bitcoindnotify/bitcoind.go:
```diff
@@ -525,30 +526,44 @@ func (b *BitcoindNotifier) confDetailsManually(confRequest chainntnfs.ConfReques
+	// where the transaction was included in.
+	batchRequest := make(map[uint32]rpcclient.FutureGetBlockHashResult, currentHeight-heightHint)
 	for height := currentHeight; height >= heightHint && height > 0; height-- {
+
```
nit: extra nl

bhandras commented on chainntnfs/bitcoindnotify/bitcoind.go:
```diff
@@ -525,30 +526,44 @@ func (b *BitcoindNotifier) confDetailsManually(confRequest chainntnfs.ConfReques
-			return nil, chainntnfs.TxNotFoundManually,
-				fmt.Errorf("unable to get hash from block "+
-					"with height %d", height)
+			return nil, chainntnfs.TxNotFoundManually, fmt.Errorf("unable to retrieve hash for "+
```
nit: formatting

bhandras commented on chainntnfs/bitcoindnotify/bitcoind.go:
```diff
@@ -525,30 +526,44 @@ func (b *BitcoindNotifier) confDetailsManually(confRequest chainntnfs.ConfReques
-			return nil, chainntnfs.TxNotFoundManually,
-				fmt.Errorf("unable to get block with hash "+
-					"%v: %v", blockHash, err)
+			return nil, chainntnfs.TxNotFoundManually, fmt.Errorf("unable to retrieve block "+
```
nit: formatting

bhandras replied:
Instead of getting the blocks one-by-one, could we also batch this?

bhandras commented on chainntnfs/bitcoindnotify/bitcoind.go:
```diff
@@ -785,19 +800,31 @@ func (b *BitcoindNotifier) historicalSpendDetails(
-	// was spent.
+	// batchRequest will store the scan requests at every height to determine
+	// if the output was spent.
+	batchRequest := make(map[uint32]rpcclient.FutureGetBlockHashResult, endHeight-startHeight)
```
nit: formatting

bhandras commented on chainntnfs/bitcoindnotify/bitcoind.go:
```diff
@@ -785,19 +800,31 @@ func (b *BitcoindNotifier) historicalSpendDetails(
 	for height := endHeight; height >= startHeight && height > 0; height-- {
-		// Ensure we haven't been requested to shut down before
-		// processing the next height.
+
```
extra nl

bhandras commented on chainntnfs/bitcoindnotify/bitcoind.go:
```diff
@@ -785,19 +800,31 @@ func (b *BitcoindNotifier) historicalSpendDetails(
+	// Sending the bulk request using updated batchClient.
+	b.chainConn.SendAsyncQueue()
+
+	for height, response := range batchRequest {
```
Again, this loop could potentially be refactored to fetch the blocks async as well.

bhandras commented on go.mod:
```diff
@@ -59,6 +59,8 @@ require (
 	gopkg.in/macaroon.v2 v2.0.0
 )
+// This replace is to leverage updated batch-json rpc
+replace github.com/btcsuite/btcwallet => ../btcwallet
```
A better way for this replace rule is to reference your pushed branch, so something like:

```
replace github.com/btcusite/btcwallet =? github.com/Vib-UX/btcwallet hash
```

Where the hash is your HEAD's hash. Then you do a `go mod tidy` and it should fetch the reference correctly. This way if someone checks out your branch they can build it without fixing these deps first.

bhandras commented on chainntnfs/btcdnotify/btcd.go:
```diff
@@ -579,31 +595,39 @@ func (b *BtcdNotifier) confDetailsManually(confRequest chainntnfs.ConfRequest,
-			return nil, chainntnfs.TxNotFoundManually,
-				fmt.Errorf("unable to get hash from block "+
-					"with height %d", height)
+			return nil, chainntnfs.TxNotFoundManually, fmt.Errorf("unable to retrieve hash for "+
```
formatting

bhandras replied:
nit: I'd just revert back the formatting here to make it fit 80 cols correctly.

bhandras commented on chainntnfs/btcdnotify/btcd.go:
```diff
@@ -579,31 +595,39 @@ func (b *BtcdNotifier) confDetailsManually(confRequest chainntnfs.ConfRequest,
-			return nil, chainntnfs.TxNotFoundManually,
-				fmt.Errorf("unable to get block with hash "+
-					"%v: %v", blockHash, err)
+			return nil, chainntnfs.TxNotFoundManually, fmt.Errorf("unable to retrieve block "+
```
formatting

bhandras replied:
nit: similarly, formatting doesn't need to change here.

bhandras commented on chainntnfs/bitcoindnotify/bitcoind_test.go:
```diff
@@ -301,3 +303,89 @@ func testHistoricalConfDetailsNoTxIndex(t *testing.T, rpcpolling bool) {
+	copy(unknownHash[:], bytes.Repeat([]byte{0x10}, 32))
+	invalidOutpoint := wire.NewOutPoint(&unknownHash, 0)
+	unknownSpendReq, err := chainntnfs.NewSpendRequest(invalidOutpoint, testScript)
+	if err != nil {
```
nit: `require.NotNil(t, err, msg)`

Vib-UX replied:
I looked few updated examples inside LND repo: https://github.com/lightningnetwork/lnd/blob/47822fada6e3766ae942cb956653bb59bf890520/chainntnfs/bitcoindnotify/bitcoind_test.go#L46

Updated it with `require.NoError(t,err,msg)`

bhandras commented on chainntnfs/bitcoindnotify/bitcoind_test.go:
```diff
@@ -301,3 +303,89 @@ func testHistoricalConfDetailsNoTxIndex(t *testing.T, rpcpolling bool) {
+		t.Fatalf("unable to retrieve historical spend details: %v", err)
+	}
+
+	assert.Equal(t, spendDetails, (*chainntnfs.SpendDetail)(nil))
```
nit: pls use `require.Equal`

bhandras commented on chainntnfs/btcdnotify/btcd.go:
```diff
@@ -58,8 +58,9 @@ type BtcdNotifier struct {
 
-	chainConn   *rpcclient.Client
-	chainParams *chaincfg.Params
+	chainConn      *rpcclient.Client
```
nit: commit message fixup idea:

```
chainntnfs: enable btcdnotify to leverage batch clients
```

bhandras replied:
Also some more detail about what is exactly changed and why could be very useful.

bhandras commented on chainntnfs/btcdnotify/btcd.go:
```diff
@@ -135,6 +136,21 @@ func New(config *rpcclient.ConnConfig, chainParams *chaincfg.Params,
+		DisableConnectOnNew:  true,
+		Certificates:         config.Certificates,
+		Endpoint:             config.Endpoint,
+		HTTPPostMode:         true,
```
Not sure but I don't think `btcd` requires http post mode?

Vib-UX replied:
![image](https://user-images.githubusercontent.com/40585900/179396986-06619e7b-0a83-465a-ab9d-8b139ca9538d.png)


When working with updated batch client `HTTPPostMode` is required.

bhandras commented on chainntnfs/btcdnotify/btcd.go:
```diff
@@ -135,6 +136,21 @@ func New(config *rpcclient.ConnConfig, chainParams *chaincfg.Params,
+		Endpoint:             config.Endpoint,
+		HTTPPostMode:         true,
+	}
+	batchChainConn, err := rpcclient.NewBatch(clientCfg)
```
nit: add extra new line above.

bhandras commented on chainntnfs/btcdnotify/btcd.go:
```diff
@@ -579,31 +595,39 @@ func (b *BtcdNotifier) confDetailsManually(confRequest chainntnfs.ConfRequest,
+	batchRequest := make(map[uint32]rpcclient.FutureGetBlockHashResult, endHeight-startHeight)
 	for height := endHeight; height >= startHeight && height > 0; height-- {
+
+		//BatchClient works with AsynRequest.
```
nit: missing space, also a more descriptive comment could be better here (or just omit).

bhandras commented on chainntnfs/bitcoindnotify/bitcoind.go:
```diff
@@ -525,47 +526,81 @@ func (b *BitcoindNotifier) confDetailsManually(confRequest chainntnfs.ConfReques
+		//BatchClient works with AsynRequest.
+		sendReq, err := b.chainConn.GetBlockHashAsync(int64(height))
+		if err != nil {
+			return nil, chainntnfs.TxNotFoundManually, fmt.Errorf("unable to "+
```
nit: pls format this to 80 cols

bhandras commented on chainntnfs/bitcoindnotify/bitcoind.go:
```diff
@@ -525,47 +526,81 @@ func (b *BitcoindNotifier) confDetailsManually(confRequest chainntnfs.ConfReques
-			return nil, chainntnfs.TxNotFoundManually,
-				fmt.Errorf("unable to get hash from block "+
-					"with height %d", height)
+			return nil, chainntnfs.TxNotFoundManually, fmt.Errorf("unable to "+
```
nit: old formatting is good, just add the err at the end (so that the change is minimal)

bhandras commented on chainntnfs/bitcoindnotify/bitcoind.go:
```diff
@@ -525,47 +526,81 @@ func (b *BitcoindNotifier) confDetailsManually(confRequest chainntnfs.ConfReques
+	index := 0
+	for height := currentHeight; height >= heightHint && height > 0; height-- {
+		// Ensure we haven't been requested to shut down before
+		// processing the next height response.
```
nit: update comment

bhandras commented on chainntnfs/bitcoindnotify/bitcoind.go:
```diff
@@ -525,47 +526,81 @@ func (b *BitcoindNotifier) confDetailsManually(confRequest chainntnfs.ConfReques
+		blockHashAddresses = append(blockHashAddresses, blockHash)
+	}
+
+	blocks, err := b.chainConn.GetBlocksBatch(blockHashAddresses)
```
It is very cool that we can now use `GetBlocksBatch` here! One thing that could break LND here is if we were to batch get say 1000 blocks. In that case low memory devices may OOM and crash LND. I think now that the main batching logic is there, you could do batch requests to get 10-25 blocks at most. This would help avoiding hogging too much memory as well as it'd be faster to quit as we'd check on the `quit` channel more often.

bhandras commented on chainntnfs/bitcoindnotify/bitcoind.go:
```diff
@@ -785,32 +820,64 @@ func (b *BitcoindNotifier) historicalSpendDetails(
+		blockHashAddresses = append(blockHashAddresses, blockHash)
+	}
+
+	blocks, err := b.chainConn.GetBlocksBatch(blockHashAddresses)
```
Similar comment to the above that we should slice this up a little to not fetch more than 10-25 blocks at once.

bhandras reviewed (commented):
Awesome progress, getting very close! Great work @Vib-UX :)

bhandras commented on chainntnfs/bitcoindnotify/bitcoind.go:
```diff
@@ -525,46 +526,102 @@ func (b *BitcoindNotifier) confDetailsManually(confRequest chainntnfs.ConfReques
 		}
 
-		blockHash, err := b.chainConn.GetBlockHash(int64(height))
+		// First, we'll fetch the block for the current height.
```
nit: not sure if this comment expresses what you meant. Maybe I'd rephrase  to something like: "Receive the next block hash from the async queue."

bhandras commented on chainntnfs/bitcoindnotify/bitcoind.go:
```diff
@@ -525,46 +526,102 @@ func (b *BitcoindNotifier) confDetailsManually(confRequest chainntnfs.ConfReques
+	}
+
+	// Now we fetch blocks in interval of 15 to avoid out of memory
+	// errors in case of fetching 1000 blocks from GetBlocksBatch().
```
nit: "in case if fetching too many blocks with GetBlocksBatch()`.

bhandras commented on chainntnfs/bitcoindnotify/bitcoind.go:
```diff
@@ -525,46 +526,102 @@ func (b *BitcoindNotifier) confDetailsManually(confRequest chainntnfs.ConfReques
+	// errors in case of fetching 1000 blocks from GetBlocksBatch().
+	intervals := len(blockHashAddresses) / 15
+
+	for i := 0; i <= intervals; i++ {
```
How about doing something like this:

```
	const batchSize = 15
	total := len(blockHashAddresses)
	for i := 0; i < total; i += batchSize {
                ...

		start := i
		end := i + batchSize

		if end > total {
			end = total
		}

                blocks, err = b.chainConn.GetBlocksBatch(
				blockHashAddresses[start:end],
		)
                ...
	}
```

bhandras replied:
Also with these changes the functionality is rather complete but I'd argue that some additional test coverage is very much needed as it's easy to add bugs when the logic is convoluted.

Vib-UX replied:
Sure we can discuss scenarios for additional test-coverages, I will try to add it accordingly

bhandras commented on chainntnfs/bitcoindnotify/bitcoind.go:
```diff
@@ -525,46 +526,102 @@ func (b *BitcoindNotifier) confDetailsManually(confRequest chainntnfs.ConfReques
+			// dispatch its confirmation details.
+			for txIndex, tx := range blocks[j].Transactions {
+				if !confRequest.MatchesTx(tx) {
+					height--
```
This seems to be incorrect to me as here we're continuing the inner for loop over the transactions. The height should only be decreased if none of the txes match. So maybe a cleaner way is to return if `confRequest.MatchesTx(tx)` otherwise decrease `height` outside before continuing the loop over the blocks.

bhandras commented on chainntnfs/bitcoindnotify/bitcoind.go:
```diff
@@ -785,52 +842,105 @@ func (b *BitcoindNotifier) historicalSpendDetails(
+	// in case of fetching 1000 blocks from GetBlocksBatch().
+	intervals := len(blockHashAddresses) / 15
+
+	for i := 0; i <= intervals; i++ {
```
Similarly as to my previous comment, you can simplify this a bit.

bhandras commented on chainntnfs/bitcoindnotify/bitcoind.go:
```diff
@@ -785,52 +842,105 @@ func (b *BitcoindNotifier) historicalSpendDetails(
+					return nil, err
+				}
+				if !matches {
+					height--
```
Again, this is incorrect, and perhaps better to return `if matches`.

Vib-UX replied:
Sure

bhandras commented on chainntnfs/bitcoindnotify/bitcoind.go:
```diff
@@ -525,46 +526,95 @@ func (b *BitcoindNotifier) confDetailsManually(confRequest chainntnfs.ConfReques
+		blocks, err := b.chainConn.GetBlocksBatch(
+			blockHashAddresses[start:end],
+		)
+
```
nit: extra nl

bhandras commented on chainntnfs/bitcoindnotify/bitcoind.go:
```diff
@@ -525,46 +526,95 @@ func (b *BitcoindNotifier) confDetailsManually(confRequest chainntnfs.ConfReques
 
-		block, err := b.GetBlock(blockHash)
-		if err != nil {
+		blockHashAddresses = append(blockHashAddresses, blockHash)
```
At this point the order of the block hashes will depend on the order of iteration of the `batchRequests` map (which is unordered so undefined). Again, testing is super important for code like this as it's easy to miss such bugs. (Same problem in `historicalSpendDetails`).

Vib-UX replied:
Yes thats a great catch, Coming from C++ background I looked if golang supports something like ordered_map but there is only unordered_map so found few alternative ways like creating slice of keys and sort the slice then traverse over the map with the sorted keys thus maintaining the order.

Here since I already know the starting point of `height` which is `currentHeight` and it will go down till `heightHint` I will traverse accordingly to maintain the order but I completely agree with basic testcoverge additional testcoverage should also be there to automatically detect these bugs.

bhandras commented on chainntnfs/bitcoindnotify/bitcoind.go:
```diff
@@ -525,8 +526,28 @@ func (b *BitcoindNotifier) confDetailsManually(confRequest chainntnfs.ConfReques
+	// Sending the bulk request using the updated batchClient.
+	b.chainConn.SendAsyncQueue()
+
+	blockHashAddresses := make([]*chainhash.Hash, 0, len(batchRequest))
```
nit: I'd rename this variable to simply `blockHashes`

Vib-UX commented on chainntnfs/bitcoindnotify/bitcoind.go:
```diff
@@ -537,34 +558,63 @@ func (b *BitcoindNotifier) confDetailsManually(confRequest chainntnfs.ConfReques
+						Tx:          tx,
+						BlockHash:   blockHashAddresses[start+j],
+						BlockHeight: uint32(height),
+						TxIndex:     uint32(txIndex),
```
nit:  If `BlockHeight: uint32(height-j),`  then on line 617 `height--`  wont be required.  similarly for `historicalSpendDetails`

Vib-UX commented on chainntnfs/bitcoindnotify/bitcoind_test.go:
```diff
@@ -221,7 +222,7 @@ func testHistoricalConfDetailsNoTxIndex(t *testing.T, rpcpolling bool) {
 	require.NoError(t, err, "unable to create conf request")
 	broadcastHeight := syncNotifierWithMiner(t, notifier, miner)
-	_, txStatus, err := notifier.historicalConfDetails(
+	txConfirmation, txStatus, err := notifier.historicalConfDetails(
```
getting txConfirmation here will increase the coverage by verifying the output results.

Vib-UX commented on chainntnfs/bitcoindnotify/bitcoind.go:
```diff
@@ -785,8 +835,26 @@ func (b *BitcoindNotifier) historicalSpendDetails(
+	// Sending the bulk request using updated batchClient.
+	b.chainConn.SendAsyncQueue()
+
+	blockHashAddresses := make([]*chainhash.Hash, 0, len(batchRequest))
```
nit: rename  to `blockhashes`

bhandras commented:
This is the latest version @Vib-UX ?

Vib-UX commented:
It's not the updated one, need to change the replace directive for the updated btcwallet and some nit fix, will push updates by tomm eod.

bhandras commented on chainntnfs/bitcoindnotify/bitcoind.go:
```diff
@@ -537,34 +558,62 @@ func (b *BitcoindNotifier) confDetailsManually(confRequest chainntnfs.ConfReques
+					return &chainntnfs.TxConfirmation{
+						Tx:          tx,
+						BlockHash:   blockHashes[start+j],
+						BlockHeight: uint32(height - j),
```
I think this is also a + ? Maybe I'm wrong, but since we iterate from `currentHeight` backwards, `height` is always the start of the current batch iuuc.

Vib-UX replied:
I think it should be -ve only, as we traverse `height` from `currentHeight` to `heightHint` with `height--`. Now since we are traversing backwards.

 Lets take an example, 
* `currentHeight =  100` 
* `start=0` 
* `batchSize = 15`  
* _j is traversing from `block=100` to `block=86`_

So here if we found the relevantTxMatch at `j=0` we return `currentHeight-j` --> _100-0_ = **100**  ✔

bhandras commented on chainntnfs/bitcoindnotify/bitcoind.go:
```diff
@@ -537,34 +558,62 @@ func (b *BitcoindNotifier) confDetailsManually(confRequest chainntnfs.ConfReques
+			return nil, chainntnfs.TxNotFoundManually, err
+		}
+
+		height := int(currentHeight) - start
```
Perhaps worth adding a comment here to explain what we track with `height`.

bhandras commented on chainntnfs/bitcoindnotify/bitcoind.go:
```diff
@@ -796,41 +863,68 @@ func (b *BitcoindNotifier) historicalSpendDetails(
+						SpenderTxHash:     &txHash,
+						SpendingTx:        tx,
+						SpenderInputIndex: inputIdx,
+						SpendingHeight:    int32(height - j),
```
Again, I believe this should be +, see other comment.

bhandras commented on chainntnfs/bitcoindnotify/bitcoind.go:
```diff
@@ -796,41 +863,68 @@ func (b *BitcoindNotifier) historicalSpendDetails(
-			}
-			if !matches {
-				continue
+		height := int(endHeight) - start
```
Please add some comment there to help readers understand the stepping.

bhandras commented on chainntnfs/bitcoindnotify/bitcoind.go:
```diff
@@ -796,41 +863,68 @@ func (b *BitcoindNotifier) historicalSpendDetails(
+				if err != nil {
+					return nil, err
+				}
+				if matches {
```
Slightly better formatting could be to:

```
if !matches {
    continue
}
```

bhandras reviewed (approved):
LGTM, but I think the indexing is off. Could be I'm wrong which also means some extra comments could be added :)

:rocket:

bhandras commented:
Ready to open against upstream once all comments are addressed. :tada:

bhandras commented on go.mod:
```diff
@@ -60,6 +60,9 @@ require (
 )
 
+// This replace is to leverage added batch-json rpc
+replace github.com/btcsuite/btcwallet => github.com/Vib-UX/btcwallet v0.14.1-0.20220806063057-36bcc1ea2655
```
Don't forget to update to the latest.

Vib-UX commented on chainntnfs/bitcoindnotify/bitcoind_test.go:
```diff
@@ -243,32 +244,120 @@ func testHistoricalConfDetailsNoTxIndex(t *testing.T, rpcpolling bool) {
+	require.Equal(t, validSpendDetails.SpentOutPoint, outpoint)
+	require.Equal(t, validSpendDetails.SpenderTxHash, spendTxHash)
+	require.Equal(t, validSpendDetails.SpendingTx, spendTx)
+	require.Equal(t, validSpendDetails.SpendingHeight, int32(broadcastHeight+1))
```
if indexing would be off this would have thrown error.

I compared `actualSpendHeight` with received BlockHeight from `validSpendDetails`.

Vib-UX committed:
bitcoindnotify: enables bitcoindnotify to use updated batch-clients.

With this RegisterSpendNtfn() & RegisterConfirmationsNtfn() services
will leverage the support of updated batchAPI for bitcoind making it
lightning fast.

Vib-UX committed:
bitcoindnotify: testcoverage for refactored bitcoindnotify.

Testcase to ensure the updated historicalSpendDetails() and confDetailsManually()
which leverages batchAPI for scanning the chain manually works as expected
with enhanced performance.
